crypto/crypto/src/x25519.rs

```diff
@@ -93,13 +96,56 @@ impl PrivateKey {
         let shared_secret = self.0.diffie_hellman(&remote_public_key);
         shared_secret.as_bytes().to_owned()
     }
+
+    /// Deserialize an X25119 PrivateKey given a canonical representation as an
+    /// bytes, following the XEdDSA approach.
+    ///
+    /// This will FAIL if the passed-in byte representation converts to a
+    /// non-canonical scalar in the X25519 sense (and thus cannot correspond to
+    /// a X25519 valid key without bit-mangling).
+    ///
+    /// This is meant to compensate for the poor key storage capabilities of
+    /// key management solutions, and NOT to promote double usage of keys under
+    /// several schemes, which would lead to BAD vulnerabilities.
+    pub fn from_scalar_bytes(scalar_slice: &[u8]) -> Result<Self, CryptoMaterialError> {
+        let ed25519_secretkey = ed25519_dalek::SecretKey::from_bytes(scalar_slice)
+            .map_err(|_| CryptoMaterialError::DeserializationError)?;
+        let expanded_key = ed25519_dalek::ExpandedSecretKey::from(&ed25519_secretkey);
+
+        let mut expanded_keypart = [0u8; 32];
+        expanded_keypart.copy_from_slice(&expanded_key.to_bytes()[..32]);
+        let potential_x25519 = x25519::PrivateKey::from(expanded_keypart);
+
+        // This checks for x25519 clamping & reduction, which is an RFC requirement
+        if &potential_x25519.to_bytes()[..] != &expanded_key.to_bytes()[..32] {
+            Err(CryptoMaterialError::DeserializationError)
+        } else {
+            Ok(potential_x25519)
+        }
+    }
 }
 
 impl PublicKey {
     /// Obtain a slice reference to the underlying bytearray
     pub fn as_slice(&self) -> &[u8] {
         &self.0
     }
+
+    /// Deserialize an X25119 PublicKey from its representation as an
+    /// Ed25519PublicKey, following the XEdDSA approach. This is meant to
+    /// compensate for the poor key storage capabilities of key management
+    /// solutions, and NOT to promote double usage of keys under several
+    /// schemes, which would lead to BAD vulnerabilities.
+    pub fn from_ed25519_bytes(ed25519_bytes: &[u8]) -> Result<Self, CryptoMaterialError> {
```

Comment:
By the name of this function, it's not very clear, what bytes these are...
May be call it: `from_ed25519_public_key_bytes` ?